File: src/status/windowtitle.go
package status

import (
	"github.com/diamondburned/gotk4/pkg/gtk/v4"
	"github.com/diamondburned/gotk4/pkg/pango"
	"github.com/javanhut/crowbar/src/hyprland"
)

// WindowTitle displays the active window title
type WindowTitle struct {
	Label  *gtk.Label
	client *hyprland.Client
}

// NewWindowTitle creates a new window title widget
func NewWindowTitle(client *hyprland.Client) *WindowTitle {
	wt := &WindowTitle{
		Label:  gtk.NewLabel(""),
		client: client,
	}

	wt.Label.AddCSSClass("window-title")
	wt.Label.SetEllipsize(pango.EllipsizeEnd)
	wt.Label.SetMaxWidthChars(50)
	wt.Label.SetHAlign(gtk.AlignStart)
	wt.Label.SetHExpand(true)

	wt.Refresh()

	return wt
}

// Refresh updates the window title display
func (wt *WindowTitle) Refresh() {
	if wt.client == nil {
		wt.Label.SetLabel("")
		return
	}

	window, err := wt.client.ActiveWindow()
	if err != nil {
		wt.Label.SetLabel("")
		return
	}

	title := window.Title
	if title == "" {
		title = window.Class
	}
	wt.Label.SetLabel(title)
}

// SetTitle sets the window title directly
func (wt *WindowTitle) SetTitle(title string) {
	wt.Label.SetLabel(title)
}

File: src/modules/window_title.rs
use crate::hyprland::HyprlandClient;
use gtk4::pango;
use gtk4::prelude::*;
use std::rc::Rc;

pub struct WindowTitle {
    pub widget: gtk4::Label,
    client: Rc<HyprlandClient>,
}

impl WindowTitle {
    pub fn new(client: Rc<HyprlandClient>) -> Self {
        let widget = gtk4::Label::new(None);
        widget.add_css_class("window-title");
        widget.set_ellipsize(pango::EllipsizeMode::End);
        widget.set_max_width_chars(50);
        widget.set_halign(gtk4::Align::Start);
        widget.set_hexpand(true);

        let wt = Self { widget, client };
        wt.refresh();
        wt
    }

    pub fn refresh(&self) {
        let Ok(window) = self.client.active_window() else {
            self.widget.set_text("");
            return;
        };

        let title = if window.title.is_empty() {
            &window.class
        } else {
            &window.title
        };
        self.widget.set_text(title);
    }
}
